src/GdbCommand.cc

```diff
@@ -31,11 +31,15 @@ static std::vector<ReplayTimeline::Mark> back_stack;
 static ReplayTimeline::Mark current_history_cp;
 static std::vector<ReplayTimeline::Mark> forward_stack;
 static SimpleGdbCommand rr_history_push(
-    "rr-history-push", [](GdbServer& gdb_server, Task*, const vector<string>&) {
+    "rr-history-push", [](GdbServer& gdb_server, Task* t, const vector<string>&) {
       if (current_history_cp) {
         back_stack.push_back(current_history_cp);
       }
-      current_history_cp = gdb_server.get_timeline().mark();
+      if (!t->session().is_replaying()) {
```

Comment:
> If you assume diversion are short-lived to perhaps evaluate a quick value, then yes you'd want no side effects.

That's all you can do with diversions right now.

If we ever change that, it will require adding new rr commands, which can take responsibility for clearing the forward stack etc.